Enable AAD access for Question Answering Service

rokulka opened the pull request:
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

rokulka committed:
Enable AAD access for QuestionAnswering

rokulka committed:
Enable AAD access for QuestionAnsweringProjectsClient

rokulka committed:
Update Tests

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/QuestionAnsweringClient.cs:
```diff
@@ -52,6 +52,30 @@ public QuestionAnsweringClient(Uri endpoint, AzureKeyCredential credential, Ques
+            options ??= new QuestionAnsweringClientOptions();
+
+            Diagnostics = new ClientDiagnostics(options, true);
+            Pipeline = HttpPipelineBuilder.Build(options, new HttpPipelinePolicy[] { new BearerTokenAuthenticationPolicy(credential, "https://cognitiveservices.azure.com/.default") }, Array.Empty<HttpPipelinePolicy>(), new ResponseClassifier());
```
Is this scope the same for all clouds, though?

rokulka replied:
Good point, other clouds would have a different suffix for the endpoint.  Is there any example that you could point to for having different scope for different clouds?

heaths replied:
Great question, and I know [I'm guilty of it as well](https://github.com/Azure/azure-sdk-for-net/blob/4ee1f303c7c246de1b5da8f5467b1d6af2f75894/sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/ConversationAnalysisClient.cs#L66).

@schaabs @tg-msft do we have any patterns today that customize the scope for the `BearerTokenAuthenticationPolicy`?

schaabs replied:
If the scope can be inferred from the endpoint you can automatically set the correct scope in the client constructor. If not you'll need to add an `Audience` property to your `ClientOptions` and provide an extensible enum with well known cloud audiences. 

You can use TextAnalytics as a guild. They added the [Audience](https://github.com/Azure/azure-sdk-for-net/blob/7091cb1413ae9639a2f620fbaceacbbd2ba014e2/sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClientOptions.cs#L59) property to `TextAnalyticsClientOptions` and added the extensible enum [TextAnalyticsAudience](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsAudience.cs).

rokulka replied:
Thanks for the suggestions, updated the PR.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/QuestionAnsweringProjectsClient.cs:
```diff
@@ -5,12 +5,37 @@
+            options ??= new QuestionAnsweringClientOptions();
+
+            ClientDiagnostics = new ClientDiagnostics(options, true);
+            _pipeline = HttpPipelineBuilder.Build(options, new HttpPipelinePolicy[] { new BearerTokenAuthenticationPolicy(credential, "https://cognitiveservices.azure.com/.default") }, Array.Empty<HttpPipelinePolicy>(), new ResponseClassifier());
```
Ditto

heaths reviewed (changes requested):
Should also add a test for AAD tokens like I did for CLU. I instrument a client specifically for this purpose. See https://github.com/Azure/azure-sdk-for-net/blob/4ee1f303c7c246de1b5da8f5467b1d6af2f75894/sdk/cognitivelanguage/Azure.AI.Language.Conversations/tests/ConversationAnalysisClientLiveTests.cs#L171-L190

rokulka committed:
Add Audience property to support multiple clouds

rokulka committed:
Update AAD tests

rokulka commented:
> Should also add a test for AAD tokens like I did for CLU. I instrument a client specifically for this purpose. See
> 
> https://github.com/Azure/azure-sdk-for-net/blob/4ee1f303c7c246de1b5da8f5467b1d6af2f75894/sdk/cognitivelanguage/Azure.AI.Language.Conversations/tests/ConversationAnalysisClientLiveTests.cs#L171-L190

Added those tests. Please take a look and verify.

rokulka committed:
Fix for AAD QuestionAnsweringClient

rokulka committed:
Fix for AAD QuestionAnsweringClient

rokulka committed:
Update Live Test Session Records

rokulka commented:
> > Should also add a test for AAD tokens like I did for CLU. I instrument a client specifically for this purpose. See
> > https://github.com/Azure/azure-sdk-for-net/blob/4ee1f303c7c246de1b5da8f5467b1d6af2f75894/sdk/cognitivelanguage/Azure.AI.Language.Conversations/tests/ConversationAnalysisClientLiveTests.cs#L171-L190
> 
> Added those tests, but I was not able to run the test and add the SessionRecords. Might need some help. Okay if we do a quick sync over a call?

I was able to run the live tests by modifying the RecordedTestMode and get the SessionRecords. I've udpated the PR.

However, I had to use my logged-in credentials via `new DefaultAzureCredential(),` instead of `TestEnvironment.Credential` while instantiating the client in the tests.

heaths commented:
> However, I had to use my credentials via `new DefaultAzureCredential(),` instead of `TestEnvironment.Credential` while instantiating the client in the tests.

@rokulka that won't be sustainable. As long as you're using the test application it should work. I'll send you a link how to query those and set them in your environment that we use for CLU.

heaths commented:
Per offline conversation, to set the variables the test credential needs, run `eng/common/TestResources/New-TestResources.ps1 cognitivelanguage` and then run your tests. See https://github.com/Azure/azure-sdk-for-net/blob/main/eng/common/TestResources/README.md for more information.

heaths committed:
Ignore AAD support tests

Not supported in the PPE environment. Need to switch over to the same static test resources for CLU. See #29958

heaths committed:
Update changelog
